cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -2837,21 +2840,30 @@ const char* soundex_utf8(gdv_int64 ctx, const char* in, gdv_int32 in_len,
   for (int i = start_idx, l = in_len; i < l; i++) {
     if (isalpha(in[i]) > 0) {
       c = toupper(in[i]) - 65;
-      if (mappings[c] != '0') {
-        if (mappings[c] != ret[si - 1]) {
-          ret[si] = mappings[c];
-          si++;
-        }
-
-        if (si > 3) break;
+      if (mappings[c] != soundex[si - 1]) {
+        soundex[si] = mappings[c];
+        si++;
       }
     }
   }
 
-  if (si <= 3) {
-    while (si <= 3) {
-      ret[si] = '0';
-      si++;
+  for (int i = 1; i < si; i++) {
+    // If the saved letter's digit is the same as the resulting first digit, remove the
+    // digit.
+    if (i == 1 && soundex[i] == mappings[ret[0] - 65]) {
```

Comment:
i == 1 is only true once. We shouldn't be checking for i==1 in the loop.

Please change this to be:
if (soundex[1] == mappings[ret[0] - 65]) {
  i = 2;
} else {
  i = 1;
}

before the for loop in line 2850

Reply:
Done